contracts/modules/cover/Cover.sol

```diff
@@ -638,64 +639,44 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard {
 
   /* ========== PRODUCT CONFIGURATION ========== */
 
-  function addProducts(
-    Product[] calldata newProducts,
-    string[] calldata ipfsMetadata
-  ) external override onlyAdvisoryBoard {
-    uint initialProductsCount = _products.length;
-
+  function setProducts(ProductParam[] calldata productParams) external override onlyAdvisoryBoard {
     uint32 _coverAssetsFallback = coverAssetsFallback;
     uint productTypesCount = _productTypes.length;
 
-    for (uint i = 0; i < newProducts.length; i++) {
-      validateProduct(newProducts[i], _coverAssetsFallback, productTypesCount);
-      _products.push(newProducts[i]);
-      emit ProductSet(initialProductsCount + i, ipfsMetadata[i]);
-    }
-  }
-
-  function addProductTypes(
-    ProductType[] calldata newProductTypes,
-    string[] calldata ipfsMetadata
-  ) external override onlyAdvisoryBoard {
-    uint initialProuctTypesCount = _productTypes.length;
-    for (uint i = 0; i < newProductTypes.length; i++) {
-      _productTypes.push(newProductTypes[i]);
-      emit ProductTypeSet(initialProuctTypesCount + i, ipfsMetadata[i]);
-    }
-  }
-
-  function editProductTypes(
-    uint[] calldata productTypeIds,
-    uint16[] calldata gracePeriodsInDays,
-    string[] calldata ipfsMetadata
-  ) external override onlyAdvisoryBoard {
-
-    for (uint i = 0; i < productTypeIds.length; i++) {
-      _productTypes[productTypeIds[i]].gracePeriodInDays = gracePeriodsInDays[i];
-      emit ProductTypeSet(productTypeIds[i], ipfsMetadata[i]);
+    for (uint i = 0; i < productParams.length; i++) {
+      ProductParam calldata param = productParams[i];
+      validateProduct(param.product, _coverAssetsFallback, productTypesCount);
+      // New product has id == uint256.max
+      if (param.productId == type(uint256).max) {
+        emit ProductSet(_products.length, param.ipfsMetadata);
+        _products.push(param.product);
+      } else {
+        // existing product
+        require(param.productId < _products.length, "Cover: Product doesnt exist. Set id to uint256.max to add it");
+        Product storage newProductValue = _products[param.productId];
+        newProductValue.isDeprecated = param.product.isDeprecated;
+        newProductValue.coverAssets = param.product.coverAssets;
+        newProductValue.initialPriceRatio = param.product.initialPriceRatio;
+        newProductValue.capacityReductionRatio = param.product.capacityReductionRatio;
+        emit ProductSet(param.productId, param.ipfsMetadata);
```

Comment:
If the product update did not intend to edit the metadata - we'll end up with a bunch of events that did not in fact change the ipfs metadata. This makes it harder for the UI to figure out when the cover metadata was in fact changed as it would have a bunch of redundant events, and also error prone because it's harder to fetch events compared to reading a product from the state at the lastest block. A transient error in event fetching could accidentally downgrade the ipfs metadata.

How about allowing the ipfs metadata to be empty and use it as a signal not to update it?

```solidity
if (param.ipfsMetadata.length > 0) {
  emit ProductSet(param.productId, param.ipfsMetadata);
}
```